contributors/design-proposals/storage/local-storage-pv.md

```diff
@@ -761,3 +761,213 @@ The central manager will poll each worker for their discovered volumes and creat
+
+It stores latest StorageClass objects in cache, they are used:
+
+* To reconcile capacity, the StorageManager need to compare actual capacity with that recorded
+in StorageClass objects
+* During volume creation, some opitions (e.g. fsType for `mkfs`) are specified in StorageClasses
+as Parameters
+
+And for PVCs, when the `populator` finds:
+
+1. An unbound PVC is annotated with `annSelectedNode`, and its node name match the value
+of the annotation
+2. Value of the `annStorageProvisioner` on the PVC matches the name of the provisioner
+
+It then trigger volume provsioning, detailed in the volume creation section.
+
+For these purposes, the following permissions are required:
+
+* list/get PersistentVolumeClaims
+* list/get/update StorageClasses
+
+##### Volume Creation
+
+During start-up of the provisioner, a new channel `provisionChan` will be created,
+and passed into both `populator` and `storageManager`:
+
+```
+provisionChan chan *v1.PersistentVolumeClaim
+```
+When it finds a PVC is in need of privsioning, as desicribed above,
+the `populator` will pass it into the channel to trigger volume privisioning;
+`StorageManager` will be the consumer, it will watch the channel and
+manage provisioning when new input received, that is:
+
+1. Create volumes behind a PV accordingly
+2. Create a PV object that is pre-bound to the PVC
+
+For LVM, the proposed workflow is:
+
+1. Check if there is already a PV pre-bound to the PVC, if so, return directly
+(This can happen when a volume has already been provisioned for a PVC,
+but has not been bound to the PVC by the controller yet).
+2. Check if Storage Class of the PVC is in record of its storageClass map.
+If not, return error.
+3. Check if the class is available for dynamic provisioning (i.e. has an input of `storageSources`).
+If not, return error.
+4. Check if there is a VG under the class, whose `Free Size` can satisfy the
+capacity requested in the PVC. If not, return error; otherwise, choose a VG for provisioning.
+5. Create `logical volume (LV)` object in the chosen VG.
+6. Initialize its file system with `mkfs`. FsType and other potential options
+for FS creation can be specified in StorageClass as Parameters
+7. Create a PV API object that is pre-bound to the PVC, and the path
+of local storage in the PV is the path of the LV created above.
+
+##### Cleanup after Release
+
+Currently we already have a component `deleter` to cleanup volumes/PVs of static local volume,
+detail cleanup actions are managed via custom scripts/binaries that specified in `blockCleanerCommand`.
+
+The behavior is extensible enough to support cleanup task of provsioned volumes too.
+So nothing new will be introduced here in the provisioner, but the difference lies
+in the cleanup script/binary itself.
+
+Existing script `quick_reset.sh` calls `mkfs -F` and `wipefs` to cleanup a block,
+For dynamic privisioned volumes, the logical volume in LVM also need to be deleted after that.

```

Comment:
Should lv deletion be done directly in the provisioner instead of a script?  Since lv creation is done directly in the provisioner?

Reply:
That makes sense, will update tp add storage manager hook inside the `deleter`